File: sdks/go/token/idtoken_test.go
package token

import (
	"crypto/rsa"
	"testing"
	"time"
)

func mintIDToken(t *testing.T, priv *rsa.PrivateKey, overrides map[string]any) string {
	t.Helper()
	iat := time.Now().Unix()
	header := map[string]any{"alg": "RS256", "typ": "JWT", "kid": "platform-1"}
	body := map[string]any{
		"iss":            "https://oidc.sudomimus.com",
		"sub":            "subject-1",
		"aud":            "client-1",
		"iat":            iat,
		"exp":            iat + 3600,
		"email":          "ada@example.com",
		"email_verified": true,
		"name":           "Ada Lovelace",
	}
	for k, v := range overrides {
		body[k] = v
	}
	return mintToken(t, header, body, priv)
}

func TestParseIDToken_ExposesBody(t *testing.T) {
	keys := generateRSAKeyPair(t)
	tok, err := ParseIDToken(mintIDToken(t, keys.privateKey, nil))
	if err != nil {
		t.Fatalf("parse: %v", err)
	}
	if tok.Body.Subject != "subject-1" || tok.Body.Email != "ada@example.com" {
		t.Fatalf("unexpected body: %+v", tok.Body)
	}
	if tok.Header.KeyID != "platform-1" {
		t.Fatalf("unexpected kid: %q", tok.Header.KeyID)
	}
}

func TestVerifyIDToken_HappyPath(t *testing.T) {
	keys := generateRSAKeyPair(t)
	jwt := mintIDToken(t, keys.privateKey, map[string]any{"nonce": "n-1"})
	tok, err := VerifyIDToken(jwt, keys.publicPEM, IDTokenExpectations{
		Audience: "client-1",
		Issuer:   "https://oidc.sudomimus.com",
		Nonce:    "n-1",
	})
	if err != nil {
		t.Fatalf("verify: %v", err)
	}
	if tok.Body.Subject != "subject-1" {
		t.Fatalf("unexpected subject: %q", tok.Body.Subject)
	}
}

func TestVerifyIDToken_Expired(t *testing.T) {
	keys := generateRSAKeyPair(t)
	past := time.Now().Unix() - 10
	jwt := mintIDToken(t, keys.privateKey, map[string]any{"iat": past - 3600, "exp": past})
	_, err := VerifyIDToken(jwt, keys.publicPEM, IDTokenExpectations{})
	assertCode(t, err, ErrExpired)
}

func TestVerifyIDToken_WrongSignature(t *testing.T) {
	keys := generateRSAKeyPair(t)
	other := generateRSAKeyPair(t)
	jwt := mintIDToken(t, keys.privateKey, nil)
	_, err := VerifyIDToken(jwt, other.publicPEM, IDTokenExpectations{})
	assertCode(t, err, ErrInvalidSignature)
}

func TestVerifyIDToken_WrongNonce(t *testing.T) {
	keys := generateRSAKeyPair(t)
	jwt := mintIDToken(t, keys.privateKey, map[string]any{"nonce": "n-1"})
	_, err := VerifyIDToken(jwt, keys.publicPEM, IDTokenExpectations{Nonce: "n-2"})
	assertCode(t, err, ErrWrongNonce)
}

File: sdks/java/token/src/test/java/com/sudomimus/token/IdTokenTest.java
package com.sudomimus.token;

import org.junit.jupiter.api.Test;

import java.security.interfaces.RSAPrivateKey;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class IdTokenTest {

    private static String mintIdToken(RSAPrivateKey priv, Map<String, Object> overrides) throws Exception {
        long iat = Instant.now().getEpochSecond();
        Map<String, Object> header = new LinkedHashMap<>();
        header.put("alg", "RS256");
        header.put("typ", "JWT");
        header.put("kid", "platform-1");

        Map<String, Object> body = new LinkedHashMap<>();
        body.put("iss", "https://oidc.sudomimus.com");
        body.put("sub", "subject-1");
        body.put("aud", "client-1");
        body.put("iat", iat);
        body.put("exp", iat + 3600);
        body.put("email", "ada@example.com");
        body.put("email_verified", true);
        body.put("name", "Ada Lovelace");
        body.putAll(overrides);

        return TestHelpers.mintToken(header, body, priv);
    }

    @Test
    void parse_exposesBody() throws Exception {
        TestHelpers.RsaKeyPair keys = TestHelpers.generateRsaKeyPair();
        IdToken token = IdToken.parse(mintIdToken(keys.privateKey, Map.of()));

        assertEquals("subject-1", token.getBody().subject);
        assertEquals("ada@example.com", token.getBody().email);
        assertEquals("platform-1", token.getHeader().keyId);
    }

    @Test
    void verify_happyPath() throws Exception {
        TestHelpers.RsaKeyPair keys = TestHelpers.generateRsaKeyPair();
        String jwt = mintIdToken(keys.privateKey, Map.of("nonce", "n-1"));

        IdToken token = IdToken.verify(jwt, keys.publicPem, new IdTokenExpectations()
                .audience("client-1")
                .issuer("https://oidc.sudomimus.com")
                .nonce("n-1"));

        assertEquals("subject-1", token.getBody().subject);
    }

    @Test
    void verify_expiredThrows() throws Exception {
        TestHelpers.RsaKeyPair keys = TestHelpers.generateRsaKeyPair();
        long past = Instant.now().getEpochSecond() - 10;
        String jwt = mintIdToken(keys.privateKey, Map.of("iat", past - 3600, "exp", past));

        TokenException ex = assertThrows(TokenException.class,
                () -> IdToken.verify(jwt, keys.publicPem, new IdTokenExpectations()));
        assertEquals(TokenErrorCode.EXPIRED, ex.getCode());
    }

    @Test
    void verify_wrongSignatureThrows() throws Exception {
        TestHelpers.RsaKeyPair keys = TestHelpers.generateRsaKeyPair();
        TestHelpers.RsaKeyPair other = TestHelpers.generateRsaKeyPair();
        String jwt = mintIdToken(keys.privateKey, Map.of());

        TokenException ex = assertThrows(TokenException.class,
                () -> IdToken.verify(jwt, other.publicPem, new IdTokenExpectations()));
        assertEquals(TokenErrorCode.INVALID_SIGNATURE, ex.getCode());
    }

    @Test
    void verify_wrongNonceThrows() throws Exception {
        TestHelpers.RsaKeyPair keys = TestHelpers.generateRsaKeyPair();
        String jwt = mintIdToken(keys.privateKey, Map.of("nonce", "n-1"));

        TokenException ex = assertThrows(TokenException.class,
                () -> IdToken.verify(jwt, keys.publicPem, new IdTokenExpectations().nonce("n-2")));
        assertEquals(TokenErrorCode.WRONG_NONCE, ex.getCode());
    }
}
